features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c

```diff
@@ -0,0 +1,333 @@
+
+
+    block_size = ctx->cipher_info->block_size;
+    if( block_size != SASI_AES_BLOCK_SIZE_IN_BYTES )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    cmac_ctx = ctx->cmac_ctx;
+
+    if( cmac_ctx->is_cc_initiated == 0 )
+    {
+        ret = init_cc( cmac_ctx );
+        if( ret != 0 )
+            goto exit;
+    }
+
+    /* Is there data still to process from the last call?
+     */
+    if( cmac_ctx->unprocessed_len > 0  )
+    {
+        const size_t size_to_copy = ilen > ( block_size - cmac_ctx->unprocessed_len ) ?
+                                    block_size - cmac_ctx->unprocessed_len : ilen;
+        memcpy( &cmac_ctx->unprocessed_block[cmac_ctx->unprocessed_len],
+                input, size_to_copy );
+        cmac_ctx->unprocessed_len += size_to_copy;
+        input += size_to_copy;
+        ilen -= size_to_copy;
+
+        /*
+         * Process the unproccessed data, in case it reached a full AES block,
+         * and there is there is still input data.
+         */
+        if( cmac_ctx->unprocessed_len == SASI_AES_BLOCK_SIZE_IN_BYTES && ilen > 0 )
+        {
+            if( SaSi_AesBlock( &cmac_ctx->CC_Context, cmac_ctx->unprocessed_block,
+                              SASI_AES_BLOCK_SIZE_IN_BYTES, NULL ) != 0 )
+            {
+                ret = MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+                goto exit;
+            }
+            cmac_ctx->unprocessed_len = 0;
+        }
+    }
+
+    if( ilen > 0 )
+    {
+        const size_t size_to_store = ( ilen % SASI_AES_BLOCK_SIZE_IN_BYTES == 0) ?
+                                     SASI_AES_BLOCK_SIZE_IN_BYTES : ilen % SASI_AES_BLOCK_SIZE_IN_BYTES;
+        memcpy( &cmac_ctx->unprocessed_block[0],
+                           input + ilen - size_to_store,
+                           size_to_store );
+        cmac_ctx->unprocessed_len = size_to_store;
+        ilen -= size_to_store;
+        if( ilen > 0 )
+        {
+            if( SaSi_AesBlock( &cmac_ctx->CC_Context, ( uint8_t * )input,
+                               ilen, NULL ) != 0 )
+            {
+                ret = MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+                goto exit;
+            }
+        }
+    }
+
+exit:
+    if( ret != 0 )
+    {
+        mbedtls_platform_zeroize( cmac_ctx, sizeof( *cmac_ctx ) );
+        mbedtls_free( cmac_ctx );
+        SaSi_AesFree( &cmac_ctx->CC_Context );
+    }
+    return( ret );
+}
+
+int mbedtls_cipher_cmac_finish( mbedtls_cipher_context_t *ctx,
+                                unsigned char *output )
+{
+    mbedtls_cmac_context_t* cmac_ctx;
+    int ret = 0;
+    size_t olen = SASI_AES_BLOCK_SIZE_IN_BYTES;
+
+    if( ctx == NULL || ctx->cipher_info == NULL ||
+        ctx->cmac_ctx == NULL || output == NULL )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    cmac_ctx = ctx->cmac_ctx;
+
+    if( cmac_ctx->is_cc_initiated == 0 )
+    {
+        ret = init_cc( cmac_ctx );
+        if( ret != 0 )
+            goto exit;
+    }
```

Comment:
Why is it that we need to call `init_cc()` here and in `mbedtls_cipher_cmac_update()` instead of calling it during initialization in `mbedtls_cipher_cmac_starts()`? It just looks to me like the program would not work without calling `mbedtls_cipher_cmac_starts()` at some point anyways because otherwise who would allocate the cmac context?

Reply:
because in the flow, the intiailization can be called only once, and some sequential cmac calls can be separated with different calls to `mbedtls_cipher_cmac_starts()`.
`init_cc()` should be called to initialize a new CC CMAC context

Reply:
I understand the first part of your answer, but not this:
```
some sequential cmac calls can be separated with different calls to mbedtls_cipher_cmac_starts().
```
what do you mean?

Reply:
I am not sure what I meant...
Anyway, I don't remember at the moment, but I believe I had a reason not call `init_cc()` from starts. I think it's because of the `reset` functionality.
But I think that in multithreading environment, we would like to set the key only before doing the actual operation.